src/coreclr/src/jit/morph.cpp

```diff
@@ -7218,306 +7203,732 @@ GenTree* Compiler::fgMorphPotentialTailCall(GenTreeCall* call)
-
-    // Check for PInvoke call types that we don't handle in codegen yet.
-    assert(!call->IsUnmanaged());
-    assert(call->IsVirtual() || (call->gtCallType != CT_INDIRECT) || (call->gtCallCookie == NULL));
+        call->fgArgInfo = nullptr;
+    }
 
-    // First move the this pointer (if any) onto the regular arg list
-    GenTree* thisPtr = NULL;
+    // Put 'this' in normal param list
     if (call->gtCallThisArg != nullptr)
     {
+        JITDUMP("Moving this pointer into arg list\n");
+        GenTree* thisPtr    = nullptr;
         GenTree* objp       = call->gtCallThisArg->GetNode();
         call->gtCallThisArg = nullptr;
 
-        if ((call->gtFlags & GTF_CALL_NULLCHECK) || call->IsVirtualVtable())
+        if (call->NeedsNullCheck())
         {
-            thisPtr      = gtClone(objp, true);
+            // clone "this" if "this" has no side effects.
+            if ((thisPtr == nullptr) && !(objp->gtFlags & GTF_SIDE_EFFECT))
+            {
+                thisPtr = gtClone(objp, true);
+            }
+
             var_types vt = objp->TypeGet();
-            if (thisPtr == NULL)
+            if (thisPtr == nullptr)
             {
-                // Too complex, so use a temp
+                // create a temp if either "this" has side effects or "this" is too complex to clone.
+
+                // tmp = "this"
                 unsigned lclNum = lvaGrabTemp(true DEBUGARG("tail call thisptr"));
                 GenTree* asg    = gtNewTempAssign(lclNum, objp);
-                if (!call->IsVirtualVtable())
-                {
-                    // Add an indirection to get the nullcheck
-                    GenTree* tmp       = gtNewLclvNode(lclNum, vt);
-                    GenTree* nullcheck = gtNewNullCheck(tmp, compCurBB);
-                    asg                = gtNewOperNode(GT_COMMA, TYP_VOID, asg, nullcheck);
-                }
-                objp    = gtNewOperNode(GT_COMMA, vt, asg, gtNewLclvNode(lclNum, vt));
-                thisPtr = gtNewLclvNode(lclNum, vt);
+
+                // COMMA(tmp = "this", deref(tmp))
+                GenTree* tmp = gtNewLclvNode(lclNum, vt);
+                GenTree* ind = gtNewOperNode(GT_IND, TYP_INT, tmp);
+                asg          = gtNewOperNode(GT_COMMA, TYP_VOID, asg, ind);
+
+                // COMMA(COMMA(tmp = "this", deref(tmp)), tmp)
+                thisPtr = gtNewOperNode(GT_COMMA, vt, asg, gtNewLclvNode(lclNum, vt));
             }
-            else if (!call->IsVirtualVtable())
+            else
             {
-                GenTree* nullcheck = gtNewNullCheck(thisPtr, compCurBB);
-                objp               = gtNewOperNode(GT_COMMA, vt, nullcheck, objp);
-                thisPtr            = gtClone(thisPtr, true);
+                // thisPtr = COMMA(deref("this"), "this")
+                GenTree* ind = gtNewOperNode(GT_IND, TYP_INT, thisPtr);
+                thisPtr      = gtNewOperNode(GT_COMMA, vt, ind, gtClone(objp, true));
```

Comment:
We now express all unconsumed indirections as `GT_NULLCHECK`.
```suggestion
                GenTree* nullcheck = gtNewNullCheck(thisPtr, compCurBB);
                thisPtr            = gtNewOperNode(GT_COMMA, vt, nullcheck, gtClone(objptr, true));	
```